### PR TITLE
Infer EOLs from existing file or use OS EOL

### PR DESCRIPTION
**Pull request by iredchuk**
**Summary**

Yarn uses UNIX line endings on all systems when writing lock file. We want to solve this problem as described in  #1061: infer line endings from the existing file, and if it does not exist, use the system default.

**Test plan**

On a UNIX system, in an empty directory:
$ yarn add <some_package>
$ cat -e yarn.lock
You should only see LF line endings.
$ unix2dos yarn.lock
$ yarn add <some_other_package>
$ cat -e yarn.lock
You should only see CRLF line endings.

On a Windows system, in an empty directory:
$ yarn add <some_package>
$ cat.exe -e yarn.lock
You should only see CRLF line endings.
$ dos2unix.exe yarn.lock
$ yarn add <some_other_package>
$ cat -e yarn.lock
You should only see LF line endings.

### CODE REVIEW

**Comment by iredchuk** [comment]
@bestander Will try, thanks.

**Comment by iredchuk** [comment]
@bestander Added, please check.

File: src/util/fs.js [hunk 1]
```diff
@@ -454,3 +455,33 @@ export async function getFileSizeOnDisk(loc: string): Promise<number> {
 export function normalizeOS(body: string): string {
   return body.replace(/\r\n/g, '\n');
 }
+
+const cr = new Buffer('\r', 'utf8')[0];
+const lf = new Buffer('\n', 'utf8')[0];
+
+async function getEolFromFile(path: string) : Promise<string | void>  {
```
Comment (bestander) [inline src/util/fs.js]: The unfortunate bit is that it adds a read file to every write file operation.
That could unexpectedly slow down quite a few things.
Any ideas how to apply this only to yarn.lock/package.json writing?

Comment (iredchuk) [inline src/util/fs.js]: IMHO putting the line ending setting into .yarnrc would be not so nice when team members work on different OSs, which is quite probable for a relatively big project.

I tried to [measure the overhead](https://gist.github.com/iredchuk/1ed83441c6985a4cbd2007b07beb273f) and got about 20ms for 10000 read iterations, so for one read it should not be significant.

However, we could limit this behaviour to only files included into source control. Is it only yarn.lock and package.json? If so, maybe we could expose both write functions and use them selectively?

Comment (bestander) [inline src/util/fs.js]: I would use this explicitly for yarn.lock, package.json and .yarnrc files only then.
How about fs would expose fs.writeFilePreservingEol and we would use it sparsely?
This kind of file parsing does not sit well with me.

Comment (iredchuk) [inline src/util/fs.js]: Makes sense. Updated.

File: src/util/fs.js [hunk 2]
```diff
@@ -454,3 +455,33 @@ export async function getFileSizeOnDisk(loc: string): Promise<number> {
...
+  let buffer;
+  try {
+    buffer = await readFileBuffer(path);
```
Comment (bestander) [inline src/util/fs.js]: better return undefined if files does not exist (check if it exists first).
And don't catch error in readFile

Comment (iredchuk) [inline src/util/fs.js]: I thought about it, but if the file is inaccessible for some other reason - is it OK to just let it fail here? Will this be handled on a higher level?

**Comment by bestander** [comment]
Commits still look quite broken in the PR.
They repeat recently merged ones

**Comment by iredchuk** [comment]
@bestander Sorry I messed up with rebasing :( Created a new PR: https://github.com/yarnpkg/yarn/pull/1911

**Pull request closed without merging**